File: 303-range-sum-query-immutable/303-range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
  this.sums = Array(nums.length)
  let sum = 0
  nums.forEach((v, i) => {
    sum += v
    this.sums[i] = sum
  })
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
  return this.sums[right] - (left === 0 ? 0 : this.sums[left-1])
}

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

File: 303-range-sum-query-immutable/303-range-sum-query-immutable.cs
public class NumArray {
  int[] sums { get; }
  public NumArray(int[] nums) {
    sums = new int[nums.Length+1];
    int sum = 0;
    for (int i = 0; i < nums.Length; i++) {
      sum += nums[i];
      sums[i+1] = sum;
    }
  }

  public int SumRange(int left, int right) {
    return sums[right + 1] - sums[left];
  }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.SumRange(left,right);
 */